validator/no_fragment_cycles.go

```diff
@@ -0,0 +1,92 @@
+package validator
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/vektah/gqlparser"
+)
+
+func init() {
+	addRule("NoFragmentCycles", func(observers *Events, addError addErrFunc) {
+		observers.OnFragment(func(walker *Walker, parentDef *gqlparser.Definition, fragment *gqlparser.FragmentDefinition) {
+			detectCycleRecursive(walker, observers, fragment, addError)
+		})
+	})
+}
+
+func detectCycleRecursive(walker *Walker, observers *Events, fragment *gqlparser.FragmentDefinition, addError addErrFunc) {
```

Comment:
It feels like this is reimplementing a big chunk of the walker. Is there a way to avoid this?

perhaps if the walker allowed the first dupe through before blocking it? then the rule can see it happen twice and only needs to listen to the events and watch for dupes?

Reply:
I transcribed [this implementation](https://github.com/graphql/graphql-js/blob/master/src/validation/rules/NoFragmentCycles.js) obediently.

> perhaps if the walker allowed the first dupe through before blocking it? then the rule can see it happen twice and only needs to listen to the events and watch for dupes?

sorry, My english is little bit bad. I can't understand what this sentences mean clearly. 😿 
Is it means "What happened if we handle the same fragment twice?" ?

Reply:
We should try to keep the implementation similar, but there are lots of things in the js version I think can be improved.

Just compare https://github.com/vektah/gqlparser/blob/master/validator/unique_directives_per_location.go
and
https://github.com/graphql/graphql-js/blob/master/src/validation/rules/UniqueDirectivesPerLocation.js

There are two different things happening, the first is the walker should never create an infinite loop even with no rules on it, so it can internally track these repeats with something like:

```golang
func (w *Walker) walkFragment(it *gqlparser.FragmentDefinition) {
	w.ActiveFragments[it.Name]++
	defer w.ActiveFragments[it.Name]-- // make sure when we finish to stop counting it

	parentDef := w.Schema.Types[it.TypeCondition.Name()]

	w.walkDirectives(parentDef, it.Directives, gqlparser.LocationFragmentDefinition)

        // still fire the triggers even though we know we have seen it before
	for _, v := range w.Observers.fragment {
		v(w, parentDef, it)
	}

 	// but dont recurse any further to protect from the loop
        if w.ActiveFragments[it.Name] == 1 {
		for _, child := range it.SelectionSet {
			w.walkSelection(parentDef, child)
		}
	}
}
```
The second is the rule, showing the right validation message, which should be much simpler to detect now that the walker is doing the tracking for us:
```golang
	addRule("NoFragmentCycles", func(observers *Events, addError addErrFunc) {
		observers.OnFragment(func(walker *Walker, parentDef *gqlparser.Definition, fragment *gqlparser.FragmentDefinition) {
			if walker.ActiveFragments[it.Name] > 1 {
				addError(...)
			}
		})
	})
```

Reply:
make sense. thanks about details!

I just added 1 commit. It's `walker should never create an infinite loop even with no rules on it` implementation.
I thought If we find a FragmentSpread with the same name, we do not need to process it twice.
because I uses `bool` instead of `int` counter.

> The second is the rule

I tried it but I can't do as well...
The problem is occurred in below test case.
https://github.com/vektah/gqlparser/blob/master/spec/validation/NoFragmentCycles.yml#L92-L104
We should emit `Cannot spread fragment "fragA" within itself via fragB.` only.
but simple implementation emit `Cannot spread fragment "fragA" within itself via fragB.` and `Cannot spread fragment "fragB" within itself via fragA.`.
If you have any good ideas, please refactoring this implementations... 😿

Reply:
Hmm, that test is a bit weird. the location of both fragment spreads are in the error message :/ And the next test is showing its order dependent.

If thats all thats failling on the simple implementation just throw a [deviation](https://github.com/vektah/gqlparser/blob/master/spec/validation/deviations.yml) in for both of those tests to return both errors

Reply:
yeah, current implementation passes both tests.
and, I'm refactoring this rule that is not depends on Walker's field.
I hope to ship this PR. We can refactoring later.